### PR TITLE
PublishTrimmed and reflection-based serialization

### LINKED ISSUE #37730: [Breaking change]: Projects with `PublishTrimmed` enabled now fail reflection-based serialization by default.
### Description

In https://github.com/dotnet/runtime/pull/88480 we made a change where projects enabling the `PublishTrimmed` msbuild property will now automatically turn off the reflection-based defaults of System.Text.Json. In other words, setting `PublishTrimmed` to true will automatically set the `JsonSerializerIsReflectionEnabledByDefault` property to false unless otherwise specified in project configuration.

### Version

.NET 8 Preview 7

### Previous behavior

Projects that have the `PublishTrimmed` property enabled, i.e.
```xml
<PropertyGroup>
  <PublishTrimmed>true</PublishTrimmed>
<PropertyGroup>
```
would publish a trimmed application, however this wouldn't necessarily disable the reflection-based default serialization behavior. In other words, the follow code:
```C#
JSonSerializer.Serialize(new { Value = 42 });
```
May or may not succeed serialization, or may or may not output the correct serialization data depending on what code got trimmed.

### New behavior

The same project will now fail serialization outright. The code:
```C#
JSonSerializer.Serialize(new { Value = 42 });
```
Now throws the exception
```console
System.InvalidOperationException: Reflection-based serialization has been disabled for this application.
```

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [X] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

We wanted to make sure that trimmed applications use appropriate defaults, and guide users towards adopting best practices suitable for trimmed applications (i.e. use the source generator and avoid accidental dependency on the unsafe reflection-based components).

### Recommended action

You can get back the original behavior by explicitly enabling the `JsonSerializerIsReflectionEnabledByDefault` property in your configuration:
```xml
<PropertyGroup>
  <PublishTrimmed>true</PublishTrimmed>
  <JsonSerializerIsReflectionEnabledByDefault>true</JsonSerializerIsReflectionEnabledByDefault>
</PropertyGroup>
```
Although we strongly recommend that users take appropriate action migrating their trimmed application to use the source generator.

### Feature area

Other (please put exact area in description textbox)

### Affected APIs

_No response_

---
[Associated WorkItem - 177283](https://dev.azure.com/msft-skilling/Content/_workitems/edit/177283)

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #37730 

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/compatibility/8.0.md](https://github.com/dotnet/docs/blob/38bc91a0485d2e9ea42a8edd5ff7dba44585670e/docs/core/compatibility/8.0.md) | [Breaking changes in .NET 8](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/8.0?branch=pr-en-us-37985) |
| [docs/core/compatibility/serialization/8.0/publishtrimmed.md](https://github.com/dotnet/docs/blob/38bc91a0485d2e9ea42a8edd5ff7dba44585670e/docs/core/compatibility/serialization/8.0/publishtrimmed.md) | [PublishedTrimmed projects fail reflection-based serialization](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/serialization/8.0/publishtrimmed?branch=pr-en-us-37985) |
| [docs/core/deploying/trimming/prepare-libraries-for-trimming.md](https://github.com/dotnet/docs/blob/38bc91a0485d2e9ea42a8edd5ff7dba44585670e/docs/core/deploying/trimming/prepare-libraries-for-trimming.md) | [Prepare .NET libraries for trimming](https://review.learn.microsoft.com/en-us/dotnet/core/deploying/trimming/prepare-libraries-for-trimming?branch=pr-en-us-37985) |

### CODE REVIEW

File: docs/core/compatibility/serialization/8.0/publishtrimmed.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+You can get back the original behavior by explicitly enabling the `JsonSerializerIsReflectionEnabledByDefault` property in your project file:
+
+```xml
+<PropertyGroup>
+  <PublishTrimmed>true</PublishTrimmed>
+  <JsonSerializerIsReflectionEnabledByDefault>true</JsonSerializerIsReflectionEnabledByDefault>
+</PropertyGroup>
+```
+
+However, we strongly recommend that you take appropriate action to migrate your trimmed applications to use the source generator.
```
Comment (gfoidl) [inline docs/core/compatibility/serialization/8.0/publishtrimmed.md:47]: Can this be put higher up in the "Recommended action" section?

So it reads like: "we disabled it for good reasons, but you can just re-enable it".
Thus I'd put the move to source generator as first point in the "Recommended action", then if that's not possible enable the reflection based serialization.

**Comment by gewarren** [comment]
@eiriktsarpalis Can you review this when you get a chance?

File: docs/core/compatibility/serialization/8.0/publishtrimmed.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+## New behavior
+
+Starting in .NET 8, projects that have the `PublishTrimmed` property enabled fail serialization outright. The code `JSonSerializer.Serialize(new { Value = 42 });` throws the following exception:
+
+> **System.InvalidOperationException: Reflection-based serialization has been disabled for this application.**
+
+## Version introduced
+
+.NET 8 Preview 7
```
Comment (eiriktsarpalis) [inline docs/core/compatibility/serialization/8.0/publishtrimmed.md:26]: I put this on the breaking change issue, but I wonder now if specifying the preview version is relevant for documentation purposes.

**Review by eiriktsarpalis — APPROVED** [review]

File: docs/core/compatibility/serialization/8.0/publishtrimmed.md (hunk 2 above)
Comment (gewarren) [reply docs/core/compatibility/serialization/8.0/publishtrimmed.md:26]: It's really only relevant before GA. I've removed the "introduced version" column from the breaking changes overview page for .NET 8, but I don't think it hurts to leave it here.

**Pull request merged**